File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        /*
        Intuition: use fasta nd slow pointers
        1. put exactly n + 1 spaces between fast and slow.
        2. Iterate until fast is pointed at nothing
        3. set slow->next = slow->next->next
        4. Return head 
        */
        if (!head) return nullptr;
        ListNode* dummy = new ListNode(-1, head);
        ListNode* slow = dummy;
        ListNode* fast = head;
        while (n > 0){
            fast = fast->next;
            n--;
        }
        while (fast){
            fast = fast->next;
            slow = slow->next;
        }
        slow->next = slow->next->next;
        return dummy->next;
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head.next == null){
            return null;
        }
        // put a fast pointer n + 1 ahead
        ListNode slow = head;
        ListNode fast = head;
        for (int i = 0; i < n + 1; i++){
            fast = fast.next;
        }
        
        // move fast to the end of the LL
        while (fast != null){
            slow = slow.next;
            fast = fast.next;
        }
        
        // slow .next = slow.next.next
        slow.next = slow.next.next;
        return head;
    }
}